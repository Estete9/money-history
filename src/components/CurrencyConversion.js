import { useParams } from 'react-router-dom';
import { useEffect } from 'react';
import { v4 as uuidv4 } from 'uuid';
import { useDispatch, useSelector } from 'react-redux';
import SectionHeader from './SectionHeader';
import { fetchCurrencyConversion } from '../redux/currencies/currenciesSlice';
import CurrencyConversionElement from './CurrencyConversionElement';
import styles from '../styles/currencyConversion.module.css';

const topCurrencyArray = ['usd', 'gbp', 'eur', 'jpy', 'chf', 'cad', 'aud', 'zar'];
const symbols = {
  eur: '€',
  usd: '$',
  zar: 'R',
  aud: 'A$',
  cad: 'C$',
  jpy: '¥',
  gbp: '£',
  chf: 'CHF',
};

function CurrencyConversion() {
  const { symbol } = useParams();
  // prettier-ignore
  const {
    currencyConversion, currenciesData, isLoadingConversion, error,
  } = useSelector((store) => store.currencies);
  const dispatch = useDispatch();
  const getCountryName = (currenciesData, symbol) => {
    const currency = currenciesData.find((currency) => currency.currencySymbol === symbol);
    return currency ? currency.currencyCountry : 'Country not found';
  };

  const countryName = getCountryName(currenciesData, symbol);

  useEffect(() => {
    if (isLoadingConversion && !currencyConversion.length) {
      dispatch(fetchCurrencyConversion({ symbol }));
    }
  }, [currencyConversion, dispatch, isLoadingConversion, symbol]);

  if (isLoadingConversion) {
    return <div>Currencies loading...</div>;
  }

  if (error) {
    return <div>{`We encountered an error: ${JSON.stringify(error)}`}</div>;
  }

  if (!currenciesData.length) {
    return (
      <section>
        <header>
          <h2>There are no currencies for the moment</h2>
          <h4>Our currency list is empty</h4>
        </header>
      </section>
    );
  }
  // prettier-ignore
  const conversionToEuro = () => (
    currencyConversion.find((conversionObj) => conversionObj.currencySymbol === 'eur').currencyConversionValue
  );

  return (
    <div>
      <SectionHeader countryName={countryName} value={conversionToEuro()} symbol={symbol} />
      <div className={styles.separator}>CONVERSION RATES</div>
      {currencyConversion.map((conversionObj) => {
        const currentSymbol = conversionObj.currencySymbol;
        return (
          topCurrencyArray.includes(currentSymbol) && (
            <div key={uuidv4()} className={styles.conversionElementWrapper}>
              <CurrencyConversionElement
                currency={{
                  currencySymbol: currentSymbol,
                  value: conversionObj.currencyConversionValue,
                  currencyCode: symbols[currentSymbol],
                }}
              />
            </div>
          )
        );
      })}
    </div>
  );
}

export default CurrencyConversion;
