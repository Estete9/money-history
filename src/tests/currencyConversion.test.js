import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import axios from 'axios';
import { Provider } from 'react-redux';
import CurrencyConversion from '../components/CurrencyConversion';

jest.mock('axios');
const mockConfigStore = configureStore([]);

describe('Integration tests for CurrencyConversion component', () => {
  test('Checks if CurrencyList renders correctly after async API consumption', async () => {
    const mockCurrenciesData = [
      {
        currencySymbol: 'AUD',
        currencyCountry: 'Australian Dollar',
      },
      {
        currencySymbol: 'CAD',
        currencyCountry: 'Canadian Dollar',
      },
      {
        currencySymbol: 'CHF',
        currencyCountry: 'Swiss Franc',
      },
      {
        currencySymbol: 'EUR',
        currencyCountry: 'Euro',
      },
      {
        currencySymbol: 'GBP',
        currencyCountry: 'British Pound Sterling',
      },
      {
        currencySymbol: 'JPY',
        currencyCountry: 'Japanese Yen',
      },
      {
        currencySymbol: 'USD',
        currencyCountry: 'United States Dollar',
      },
      {
        currencySymbol: 'ZAR',
        currencyCountry: 'South African Rand',
      },
    ];
    const mockConversionData = [
      {
        success: true,
        timestamp: 1697910064,
        historical: true,
        base: 'EUR',
        date: '2023-10-21',
        rates: {
          AUD: 1.679069,
        },
      },
      {
        success: true,
        timestamp: 1697910064,
        historical: true,
        base: 'EUR',
        date: '2020-10-21',
        rates: {
          AUD: 1.679069,
        },
      },
    ];

    axios.get.mockResolvedValue({ data: mockConversionData });

    const initialState = {
      currencies: {
        currencyConversion: mockConversionData,
        currenciesData: mockCurrenciesData,
        isLoadingConversion: false,
      },
    };

    const conversionStore = mockConfigStore(initialState);

    let component;
    await renderer.act(async () => {
      component = renderer.create(
        <Provider store={conversionStore}>
          <MemoryRouter initialEntries={['/currency/AUD']}>
            <Routes>
              <Route path="/currency/:symbol" element={<CurrencyConversion />} />
            </Routes>
          </MemoryRouter>
        </Provider>,
      );
    });
    const currencyConversionTree = component.toJSON();
    await waitFor(() => expect(currencyConversionTree).toMatchSnapshot());
  });

  test('Checks if loading screen in CurrencyConversion renders correctly', () => {
    const conversionStore = mockConfigStore({
      currencies: {
        currencyConversion: [],
        isLoadingConversion: true,
      },
    });
    const component = renderer.create(
      <Provider store={conversionStore}>
        <CurrencyConversion />
      </Provider>,
    );
    const currencyConversionTree = component.toJSON();

    expect(currencyConversionTree).toMatchSnapshot();
  });
});
