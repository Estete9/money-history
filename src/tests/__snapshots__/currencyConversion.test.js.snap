// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration tests for CurrencyConversion component Checks if CurrencyList renders correctly after async API consumption 1`] = `
<div>
  <section
    className="sectionHeader"
  >
    <div
      className="sectionHeaderImgWrapper"
    >
      <img
        alt="currency icon"
        className="sectionHeaderImg"
        src="svg-mock"
      />
    </div>
    <div
      className="sectionHeaderContentWrapper"
    >
      <h1>
        undefined 1.679069
      </h1>
      <p>
        Australian Dollar
      </p>
      <p>
        â‚¬1 euro
      </p>
    </div>
  </section>
  <div
    className="separator"
  >
    PAST 5 YEARS
  </div>
  <div
    className="conversionElementWrapper"
  >
    <div
      className="currencyConversionElementWrapper"
    >
      <div>
        2023
      </div>
      <p
        className="currencyConversion"
      >
        AUD 1.679069 = eur 1
      </p>
      <img
        alt="enter currency history button"
        className="historyArrowImg"
        src="svg-mock"
      />
    </div>
  </div>
  <div
    className="conversionElementWrapper"
  >
    <div
      className="currencyConversionElementWrapper"
    >
      <div>
        2020
      </div>
      <p
        className="currencyConversion"
      >
        AUD 1.679069 = eur 1
      </p>
      <img
        alt="enter currency history button"
        className="historyArrowImg"
        src="svg-mock"
      />
    </div>
  </div>
</div>
`;

exports[`Integration tests for CurrencyConversion component Checks if loading screen in CurrencyConversion renders correctly 1`] = `
<div>
  Currency's Conversion loading...
</div>
`;
